server/block/cactus.go

```diff
@@ -75,6 +61,34 @@ func (c Cactus) RandomTick(pos cube.Pos, w *world.World, r *rand.Rand) {
 	w.SetBlock(pos, Cactus{Age: c.Age}, nil)
 }
 
+// CanGrowHere implements logic to check if cactus can live/grow here
+func (c Cactus) CanGrowHere(pos cube.Pos, w *world.World) bool {
+	// check surroundings
+	for _, face := range cube.HorizontalFaces() {
+		if _, ok := w.Block(pos.Side(face)).(Air); !ok {
+			return false
+		}
+	}
+	// placed on sand.
+	if supportsVegetation(c, w.Block(pos.Sub(cube.Pos{0, 1}))) {
+		return true
+	}
+
+	// placed on one cactus
+	_, one := w.Block(pos.Side(cube.FaceDown)).(Cactus)
+	if one && supportsVegetation(c, w.Block(pos.Sub(cube.Pos{0, 2}))) {
+		return true
+	}
+
+	// placed on two cactii
+	_, two := w.Block(pos.Side(cube.FaceDown)).(Cactus)
+	if one && two && supportsVegetation(c, w.Block(pos.Sub(cube.Pos{0, 3}))) {
+		return true
+	}
```

Comment:
I do not think this is a particularly great idea, why not just do one.CanGrowHere() and allow it to recursively check all the cactus?

Reply:
changed to a recursive function. limiting the recursion of the height of the plant. (same idea used for sugar cane and cactus.)